File: data-structures/binary-search-tree.rb

class Node
  attr_accessor :right_child, :left_child
  attr_reader :val
  def initialize(val)
    @val = val
    @right_child = nil
    @left_child = nil
  end
end

class BinarySearchTree
  attr_accessor :root, :size
  def initialize
    @root = nil
    @size = 0
  end

  def insert(val)
    if @root === nil
      @root = Node.new(val)
      @size += 1
    else
      find_and_insert(@root, val)
      @size += 1
    end
  end

  def find_and_insert(current_node, val)
    if val > current_node.val
      if current_node.right_child === nil
        current_node.right_child = Node.new(val)
      else
        find_and_insert(current_node.right_child, val)
      end
    elsif val < current_node.val
      if current_node.left_child === nil
        current_node.left_child = Node.new(val)
      else
        find_and_insert(current_node.left_child, val)
      end
    end
  end

  def search(target)
    traverse(@root, target) == true ? true : false
  end

  def traverse(current_node, target)
    if current_node.nil?
      return
    elsif current_node.val == target
      return true
    end

    if target > current_node.val
      traverse(current_node.right_child, target)
    elsif target < current_node.val
      traverse(current_node.left_child, target)
    end
  end

  def delete(val)
    current_vals = round_up(@root, val, Array.new)
    p "Delete value does not exist" if current_vals.length == @size

    @root = nil
    @size = 0

    current_vals.each do |val|
      insert(val)
    end
  end

  def round_up(current_node, delete_val, temp)
    return if current_node.nil?
    if current_node.val != delete_val
      temp.push(current_node.val)
    end

    round_up(current_node.right_child, delete_val, temp)
    round_up(current_node.left_child, delete_val, temp)
    return temp
  end
end

binary_search_tree = BinarySearchTree.new
binary_search_tree.insert(5)
binary_search_tree.insert(6)
binary_search_tree.insert(7)
# p binary_search_tree.search(5)
# p binary_search_tree.search(3)
p binary_search_tree.delete(5)
# p binary_search_tree.delete(8)
p binary_search_tree

















# //////////////////////////////////////////////////////////
# ///////////////  DO NOT TOUCH TEST BELOW!!!  /////////////
# //////////////////////////////////////////////////////////

# require 'test/unit'

# class BinarySearchTreeNodeTest < Test::Unit::TestCase
#   def test_creation_of_node
#     test = Node.new(5)

#     assert_not_equal(nil, test)
#   end
#   def test_encoding_a_value
#     test = Node.new(5)

#     assert_equal(5, test.val)
#     assert_equal(nil, test.right_child)
#     assert_equal(nil, test.left_child)
#   end
#   def test_inserting_other_values
#     test = Node.new(5)
#     testRight = Node.new(10)
#     testLeft = Node.new(2)
#     test.right_child = testRight
#     test.left_child = testLeft

#     assert_equal(5, test.root.val)
#     assert_equal(2, test.root.left_child.val)
#     assert_equal(10, test.root.right_child.val)

#   end
# end

# class BinarySearchTreeClassTest < Test::Unit::TestCase
#   def test_Binary_Search_Tree_properties_existence
#     test = Binary_Search_Tree.new()

#     assert_respond_to(test, :root)
#     assert_respond_to(test, :size)

#   end

#   def test_Binary_Search_Tree_methods_existence
#     test = Binary_Search_Tree.new()

#     assert_respond_to(test, :insert)
#     assert_respond_to(test, :contains)
#     assert_respond_to(test, :delete)
#   end

#   def test_Binary_Search_Tree_insert_method_single_node
#     test = Binary_Search_Tree.new()
#     test.insert(5)

#     assert_equal(5, test.root.val)
#   end

#   def test_Binary_Search_Tree_insert_method_three_nodes
#     test = Binary_Search_Tree.new()
#     test.insert(5)
#     test.insert(10)
#     test.insert(2)

#     assert_equal(5, test.root.val)
#     assert_equal(10, test.root.right_child.val)
#     assert_equal(2, test.root.left_child.val)
#   end

#   def test_Binary_Search_Tree_search_method_node_exists
#     test = Binary_Search_Tree.new()
#     test.insert(5)
#     test.insert(10)
#     test.insert(2)

#     assert_equal(true, test.search(10))
#   end

#   def test_Binary_Search_Tree_search_method_node_does_not_exist
#     test = Binary_Search_Tree.new()
#     test.insert(5)
#     test.insert(10)
#     test.insert(2)

#     assert_equal(false, test.search(7))
#   end

#   def test_Binary_Search_Tree_delete_method_removes_node
#     test = Binary_Search_Tree.new()
#     test.insert(5)
#     test.insert(10)
#     test.insert(2)

#     assert_equal(true, test.search(10))

#     test.delete(10)

#     assert_equal(false, test.search(10))
#   end

#   def test_Binary_Search_Tree_delete_method_removes_node
#     test = Binary_Search_Tree.new()
#     test.insert(5)
#     test.insert(10)
#     test.insert(2)

#     assert_equal(3, test.size)

#     test.delete(10)

#     assert_equal(2, test.size)
#   end
# end

File: data-structures/binary-search-tree.js
var node = function(value){
  this.value = value;
  this.rightChild = null;
  this.leftChild = null;
}


var binarySearchTree = function(){
  this.root = null;
  this.size = 0;
}

binarySearchTree.prototype.insert = function(value) {
  if (this.root === null) {
    this.root = new node(value);
    this.size++;
  } else {
    var findAndInsert = function(currentNode) {
      if (value > currentNode.value) {
        if (currentNode.rightChild === null) {
          currentNode.rightChild = new node(value);
        } else {
          findAndInsert(currentNode.rightChild);
        }
      } else if (value < currentNode.value) {
        if (currentNode.leftChild === null) {
          currentNode.leftChild = new node(value);
        } else {
          findAndInsert(currentNode.leftChild);
        }
      }
    }
    findAndInsert(this.root);
    this.size++;
  }
}

binarySearchTree.prototype.search = function(target) {
  var check = false;

  var traverse = function(currentNode) {
    if (currentNode === null) {
      return;
    } else if (currentNode.value == target) {
      check = true;
      return;
    }

    if (target > currentNode.value) {
      traverse(currentNode.rightChild);
    } else if (target < currentNode.value) {
      traverse(currentNode.leftChild);
    }
  }

  traverse(this.root);
  return check;
}


binarySearchTree.prototype.delete = function(deleteValue) {
  var temp = [];

  var roundUp = function(currentNode) {
    if (currentNode === null) {
      return;
    }
    if (currentNode.value !== deleteValue) {
      temp.push(currentNode.value);
    }

    roundUp(currentNode.rightChild);
    roundUp(currentNode.leftChild);
  }

  roundUp(this.root);

  if (temp.length === this.size) {
    console.log('deleteValue does not exits');
    return;
  }

  this.root = null;
  this.size = 0;
  var toInsert;
  for (var i=0;i<temp.length;i++) {
    toInsert = temp[i];
    this.insert(toInsert);
  }

  console.log('deleteValue has been removed');
}





////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

var expect = require('chai').expect;

describe('binary search tree node', function(){

  describe('creation of a node', function(){
    it('should exist: ', function(){
      var test = new node();

      expect(test).to.not.equal(undefined);
    });
  });

  describe('encoding a value', function(){
    it('should store a value: ', function(){
      var test = new node();

      expect(test).to.have.property('value');
      expect(test.value).to.equal(undefined);

      test.value = 3;

      expect(test.value).to.equal(3);
    });
  });

  describe('pointing to another node', function(){
    it('should be able to point to another node: ', function(){
      var initial = new node(5);
      var rightTarget = new node(10);
      var leftTarget = new node(2);

      expect(initial).to.have.property('value');
      expect(initial).to.have.property('rightChild');
      expect(initial).to.have.property('leftChild');
      expect(initial.rightChild).to.equal(null);
      expect(initial.leftChild).to.equal(null);

      initial.rightChild = rightTarget;
      initial.leftChild = leftTarget;

      expect(initial.rightChild.value).to.equal(10);
      expect(initial.leftChild.value).to.equal(2);

    })
  })
})

describe('binary search tree class ', function(){

  describe('binarySearchTree properties', function(){
    it('should have properties root and size', function(){
      var test = new binarySearchTree();

      expect(test).to.have.property('root');
      expect(test).to.have.property('size');
      expect(test.root).to.equal(null);
      expect(test.size).to.equal(0);
    });
  });

  describe('binarySearchTree methods existence', function(){
    it('should have methods insert, search, and delete', function(){
      var test = new binarySearchTree();

      expect(test).to.respondTo('insert');
      expect(test).to.respondTo('search');
      expect(test).to.respondTo('delete');
    });
  });

  describe('binarySearchTree insert method', function(){
    it('should be able to insert a single node', function(){
      var test = new binarySearchTree();

      expect(test.root).to.equal(null);
      expect(test.size).to.equal(0);

      test.insert(5);

      expect(test.root.value).to.equal(5);
      expect(test.root.rightChild).to.equal(null);
      expect(test.root.leftChild).to.equal(null);
      expect(test.size).to.equal(1);
    });

    it('should be able to insert a second node', function(){
      var test = new binarySearchTree();

      expect(test.root).to.equal(null);
      expect(test.size).to.equal(0);

      test.insert(5);

      expect(test.root.value).to.equal(5);
      expect(test.root.rightChild).to.equal(null);
      expect(test.root.leftChild).to.equal(null);
      expect(test.size).to.equal(1);

      test.insert(10);

      expect(test.root.value).to.equal(5);
      expect(test.root.rightChild.value).to.equal(10);
      expect(test.root.leftChild).to.equal(null);
      expect(test.size).to.equal(2);
    });
  });

  describe('binarySearchTree search method', function(){
    it('should return true when the node exists', function(){
      var test = new binarySearchTree();
      test.insert(5);
      test.insert(10);

      expect(test.search(5)).to.equal(true);
    })

    it('should return false when the node does not exist', function(){
      var test = new binarySearchTree();
      test.insert(5);
      test.insert(10);

      expect(test.search(17)).to.equal(false);
    })
  })

  describe('binarySearchTree delete method', function(){
    it('should delete a node', function(){
      var test = new binarySearchTree();
      test.insert(5);
      test.insert(10);
      test.insert(2);
      test.insert(7);
      test.insert(1);

      expect(test.search(7)).to.equal(true);

      test.delete(7);

      expect(test.search(7)).to.equal(false);

    });

    it('should modify the size when deleting a node', function(){
      var test = new binarySearchTree();
      test.insert(5);
      test.insert(10);
      test.insert(2);
      test.insert(7);
      test.insert(1);

      expect(test.size).to.equal(5);

      test.delete(7);

      expect(test.size).to.equal(4);
    });

  });

})
